lang/java/reef-runtime-yarn/src/main/java/org/apache/reef/runtime/yarn/driver/YarnDriverRestartManager.java

```diff
@@ -0,0 +1,169 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.runtime.yarn.driver;
+
+import org.apache.hadoop.yarn.api.records.Container;
+import org.apache.reef.annotations.Unstable;
+import org.apache.reef.annotations.audience.DriverSide;
+import org.apache.reef.annotations.audience.Private;
+import org.apache.reef.annotations.audience.RuntimeAuthor;
+import org.apache.reef.driver.restart.DriverRestartManager;
+import org.apache.reef.proto.ReefServiceProtos;
+import org.apache.reef.runtime.common.driver.DriverStatusManager;
+import org.apache.reef.runtime.common.driver.EvaluatorPreserver;
+import org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEventImpl;
+import org.apache.reef.runtime.yarn.driver.parameters.YarnEvaluatorPreserver;
+import org.apache.reef.driver.restart.RestartEvaluatorInfo;
+import org.apache.reef.tang.annotations.Parameter;
+
+import javax.inject.Inject;
+import java.util.*;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The implementation of restart manager for YARN. Handles evaluator preservation as well
+ * as evaluator recovery on YARN.
+ */
+@DriverSide
+@RuntimeAuthor
+@Private
+@Unstable
+public final class YarnDriverRestartManager implements DriverRestartManager {
+
+  private static final Logger LOG = Logger.getLogger(YarnDriverRestartManager.class.getName());
+
+  private final EvaluatorPreserver evaluatorPreserver;
+  private final ApplicationMasterRegistration registration;
+  private final DriverStatusManager driverStatusManager;
+  private final REEFEventHandlers reefEventHandlers;
+  private List<Container> previousContainers;
+
+  @Inject
+  private YarnDriverRestartManager(@Parameter(YarnEvaluatorPreserver.class)
+                           final EvaluatorPreserver evaluatorPreserver,
+                           final REEFEventHandlers reefEventHandlers,
+                           final ApplicationMasterRegistration registration,
+                           final DriverStatusManager driverStatusManager){
+    this.registration = registration;
+    this.evaluatorPreserver = evaluatorPreserver;
+    this.driverStatusManager = driverStatusManager;
+    this.reefEventHandlers = reefEventHandlers;
+    this.previousContainers = null;
+  }
+
+  @Override
+  public boolean isRestart() {
```

Comment:
This assumes that YARN hasn't forgotten about our prior allocations, right? IIRC, this could be an issue if the RM also went through a restart. Maybe we should base this decision on other, more reliable information like the environment variable?